algorithm-simulator/src/main/scala/style95/Container.scala

```diff
@@ -24,24 +24,41 @@ object Container {
   final case class ContainerProperty(initialDelay: FiniteDuration,
                                      execTime: FiniteDuration)
 
+  // a message to instrument a graceful shutdown
+  final case object StopContainer
+
+  final case object ContainerStopped
+
+  private case object DelayKey
+
+  private case class DelayMsg(msg: ActivationMessage)
+
   def props(simulator: ActorRef, property: ContainerProperty): Props =
     Props(new Container(simulator, property))
 }
 
 class Container(val simulator: ActorRef, property: ContainerProperty)
-    extends Actor {
+    extends Actor
+    with Timers {
 
   import Container._
   import context.{system, dispatcher}
 
+  private var tombStone = false
+
   system.scheduler.scheduleOnce(property.initialDelay) {
     simulator ! ContainerCreated
   }
 
   override def receive: Receive = {
     case msg: ActivationMessage =>
-      system.scheduler.scheduleOnce(property.execTime) {
-        simulator ! WorkDone(msg)
+      timers.startSingleTimer(DelayKey, DelayMsg(msg), property.execTime)
```

Comment:
The simulator will only send `ActivationMessage` to an idle `Container`. If a `Container` is waiting for the `DelayMsg`, it will not receive anther `ActivationMessage`; thus, there will always be at most one timer.